README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# graphcoloring

[![Travis build
status](https://travis-ci.org/saurfang/graphcoloring.svg?branch=master)](https://travis-ci.org/saurfang/graphcoloring)
[![CRAN
status](https://www.r-pkg.org/badges/version/graphcoloring)](https://cran.r-project.org/package=graphcoloring)
[![lifecycle](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://www.tidyverse.org/lifecycle/#experimental)
[![Coverage
status](https://codecov.io/gh/saurfang/graphcoloring/branch/master/graph/badge.svg)](https://codecov.io/github/saurfang/graphcoloring?branch=master)

`graphcoloring` is a collection of graph coloring algorithms for
coloring vertices of a graph such that no two adjacent vertices share
the same color. The algorithms are included via the embedded
‘GraphColoring’ C++ library,
<https://github.com/brrcrites/GraphColoring>.

## Installation

You can install the released version of graphcoloring from
[CRAN](https://CRAN.R-project.org) with:

``` r
install.packages("graphcoloring")
```

Development version can be installed with

``` r
devtools::install_github("saurfang/graphcoloring")
```

## Example

`color_*` functions operate under `tidygraph` family and can be used to
color nodes within `mutate` context similar to `group_*` functions in
`tidygraph`.

``` r
library(graphcoloring)
library(tidygraph)
library(ggraph)

set.seed(42)

play_islands(5, 10, 0.8, 3) %>%
  mutate(color = as.factor(color_dsatur())) %>%
  ggraph(., layout = 'kk') +
  geom_edge_link(aes(alpha = ..index..), show.legend = FALSE) +
  geom_node_point(aes(color = color), size = 7) +
  theme_graph()
```

<img src="man/figures/README-example-1.png" width="100%" />

`graph_coloring_*` functions directly take adjacency lists and returns
an integer vector of assigned labels. For example, this can be used with
`sf::st_intersects()` to color a feature collection for visualization.

``` r
library(graphcoloring)
library(USAboundaries)
library(sf)
library(ggplot2)

set.seed(48)

us_states() %>%  
  filter(!(name %in% c("Alaska", "District of Columbia", "Hawaii", "Puerto Rico"))) %>%
  mutate(
    color = st_intersects(.) %>%
      graph_coloring_dsatur() %>%
      as.factor()
  ) %>%
  ggplot() +
  geom_sf(aes(fill = color)) +
  theme_bw()
```

<img src="man/figures/README-sf-example-1.png" width="100%" />


NEWS.md
# graphcoloring 0.1.0

* First release of graphcoloring.


tests/testthat.R
library(testthat)
library(graphcoloring)

test_check("graphcoloring")


tests/testthat/test-color.R
context("test-color")

library(tidygraph)

expect_graph_colored <- function(graph) {
  graph %>%
    activate(edges) %>%
    mutate(
      from_color = .N()$color[from],
      to_color = .N()$color[to]
    ) %>%
    filter(from_color == to_color) %>%
    as_tibble() %>%
    nrow() %>%
    expect_equal(0L)
}

test_that("graph can be colored with DSATUR", {
  expect_graph_colored(
    play_islands(5, 10, 0.8, 3) %>%
      mutate(color = as.factor(color_dsatur()))
  )
})

test_that("graph can be colored with MSC", {
  expect_graph_colored(
    play_islands(5, 10, 0.8, 3) %>%
      mutate(color = as.factor(color_msc()))
  )
})

test_that("graph can be colored with lmXRLF", {
  expect_graph_colored(
    play_islands(5, 10, 0.8, 3) %>%
      mutate(color = as.factor(color_lmxrlf()))
  )
})

test_that("bipartite graph can be colored with TabuCol using two colors only", {
  set.seed(324)
  expect_graph_colored(
    play_bipartite(8, 8, 0.4) %>%
      mutate(color = as.factor(color_tabucol(2)))
  )
})

test_that("TabuCol may fail to color a bipartite graph using two colors when iterations are insufficient", {
  set.seed(1023)
  expect_error(
    play_bipartite(8, 8, 0.4) %>%
      mutate(color = as.factor(color_tabucol(2, rep = 10, nbmax = 10)))
  )
})

test_that("TabuCol errors out when coloring is impossible", {
  expect_error(
    play_bipartite(5, 5, 0.4) %>%
      mutate(color = as.factor(color_tabucol(1))),
    "Graph cannot be colored with 1 colors!"
  )
})

test_that("graph can be colored with Hybrid DSATUR/TabuCol", {
  expect_graph_colored(
    play_islands(2, 5, 0.8, 3) %>%
      mutate(color = as.factor(color_hybrid_dsatur_tabucol()))
  )
})

# test_that("graph can be colored with Hybrid lmXRLF/TabuCol", {
#   expect_graph_colored(
#     play_islands(2, 5, 0.8, 3) %>%
#       mutate(color = as.factor(color_hybrid_lmxrlf_tabucol()))
#   )
# })

test_that("color_with expects tidygraph nodes", {
  expect_error(
    play_islands(5, 10, 0.8, 3) %>%
      activate(edges) %>%
      mutate(color = as.factor(color_dsatur())),
    "This call requires nodes to be active"
  )
})


R/RcppExports.R
# Generated by using Rcpp::compileAttributes() -> do not edit by hand
# Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393

#' @describeIn graph_coloring Color graph using DSATUR algorithm
#' \insertCite{Brelaz:1979:NMC:359094.359101}{graphcoloring}
#' @export
graph_coloring_dsatur <- function(adj_list) {
    .Call(`_graphcoloring_graph_coloring_dsatur`, adj_list)
}

#' @describeIn graph_coloring Color graph using Maximum Cardinality Search(MCS) algorithm
#' \insertCite{Palsberg:2007:RAV:1273694.1273695}{graphcoloring}
#' @export
graph_coloring_msc <- function(adj_list) {
    .Call(`_graphcoloring_graph_coloring_msc`, adj_list)
}

#' @describeIn graph_coloring Color graph using Least-constraining Most-constrained eXtended RLF(lmXRLF) algorithm
#' \insertCite{Kirovski:1998:ECL:277044.277165}{graphcoloring}
#' @export
graph_coloring_lmxrlf <- function(adj_list) {
    .Call(`_graphcoloring_graph_coloring_lmxrlf`, adj_list)
}

#' @describeIn graph_coloring Color graph using a hybrid of DASTUR and TabuCol algorithm
#' \insertCite{Kirovski:1998:ECL:277044.277165,Brelaz:1979:NMC:359094.359101,Hertz:1987:UTS:44141.44146}{graphcoloring}
#' @export
graph_coloring_hybrid_dsatur_tabucol <- function(adj_list) {
    .Call(`_graphcoloring_graph_coloring_hybrid_dsatur_tabucol`, adj_list)
}

#' @describeIn graph_coloring Color graph using a hybrid of lmXRLF and TabuCol algorithm
#' \insertCite{Kirovski:1998:ECL:277044.277165,Hertz:1987:UTS:44141.44146}{graphcoloring}
#' @export
graph_coloring_hybrid_lmxrlf_tabucol <- function(adj_list) {
    .Call(`_graphcoloring_graph_coloring_hybrid_lmxrlf_tabucol`, adj_list)
}

#' @describeIn graph_coloring Color graph using TabuCol algorithm
#' \insertCite{Hertz:1987:UTS:44141.44146}{graphcoloring}
#' @export
graph_coloring_tabucol <- function(adj_list, k, tabu_size = 25L, rep = 100L, nbmax = 1000L) {
    .Call(`_graphcoloring_graph_coloring_tabucol`, adj_list, k, tabu_size, rep, nbmax)
}



src/RcppExports.cpp
// Generated by using Rcpp::compileAttributes() -> do not edit by hand
// Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393

#include <Rcpp.h>

using namespace Rcpp;

// graph_coloring_dsatur
IntegerVector graph_coloring_dsatur(ListOf<IntegerVector> adj_list);
RcppExport SEXP _graphcoloring_graph_coloring_dsatur(SEXP adj_listSEXP) {
BEGIN_RCPP
    Rcpp::RObject rcpp_result_gen;
    Rcpp::RNGScope rcpp_rngScope_gen;
    Rcpp::traits::input_parameter< ListOf<IntegerVector> >::type adj_list(adj_listSEXP);
    rcpp_result_gen = Rcpp::wrap(graph_coloring_dsatur(adj_list));
    return rcpp_result_gen;
END_RCPP
}
// graph_coloring_msc
IntegerVector graph_coloring_msc(ListOf<IntegerVector> adj_list);
RcppExport SEXP _graphcoloring_graph_coloring_msc(SEXP adj_listSEXP) {
BEGIN_RCPP
    Rcpp::RObject rcpp_result_gen;
    Rcpp::RNGScope rcpp_rngScope_gen;
    Rcpp::traits::input_parameter< ListOf<IntegerVector> >::type adj_list(adj_listSEXP);
    rcpp_result_gen = Rcpp::wrap(graph_coloring_msc(adj_list));
    return rcpp_result_gen;
END_RCPP
}
// graph_coloring_lmxrlf
IntegerVector graph_coloring_lmxrlf(ListOf<IntegerVector> adj_list);
RcppExport SEXP _graphcoloring_graph_coloring_lmxrlf(SEXP adj_listSEXP) {
BEGIN_RCPP
    Rcpp::RObject rcpp_result_gen;
    Rcpp::RNGScope rcpp_rngScope_gen;
    Rcpp::traits::input_parameter< ListOf<IntegerVector> >::type adj_list(adj_listSEXP);
    rcpp_result_gen = Rcpp::wrap(graph_coloring_lmxrlf(adj_list));
    return rcpp_result_gen;
END_RCPP
}
// graph_coloring_hybrid_dsatur_tabucol
IntegerVector graph_coloring_hybrid_dsatur_tabucol(ListOf<IntegerVector> adj_list);
RcppExport SEXP _graphcoloring_graph_coloring_hybrid_dsatur_tabucol(SEXP adj_listSEXP) {
BEGIN_RCPP
    Rcpp::RObject rcpp_result_gen;
    Rcpp::RNGScope rcpp_rngScope_gen;
    Rcpp::traits::input_parameter< ListOf<IntegerVector> >::type adj_list(adj_listSEXP);
    rcpp_result_gen = Rcpp::wrap(graph_coloring_hybrid_dsatur_tabucol(adj_list));
    return rcpp_result_gen;
END_RCPP
}
// graph_coloring_hybrid_lmxrlf_tabucol
IntegerVector graph_coloring_hybrid_lmxrlf_tabucol(ListOf<IntegerVector> adj_list);
RcppExport SEXP _graphcoloring_graph_coloring_hybrid_lmxrlf_tabucol(SEXP adj_listSEXP) {
BEGIN_RCPP
    Rcpp::RObject rcpp_result_gen;
    Rcpp::RNGScope rcpp_rngScope_gen;
    Rcpp::traits::input_parameter< ListOf<IntegerVector> >::type adj_list(adj_listSEXP);
    rcpp_result_gen = Rcpp::wrap(graph_coloring_hybrid_lmxrlf_tabucol(adj_list));
    return rcpp_result_gen;
END_RCPP
}
// graph_coloring_tabucol
IntegerVector graph_coloring_tabucol(ListOf<IntegerVector> adj_list, int k, int tabu_size, int rep, int nbmax);
RcppExport SEXP _graphcoloring_graph_coloring_tabucol(SEXP adj_listSEXP, SEXP kSEXP, SEXP tabu_sizeSEXP, SEXP repSEXP, SEXP nbmaxSEXP) {
BEGIN_RCPP
    Rcpp::RObject rcpp_result_gen;
    Rcpp::RNGScope rcpp_rngScope_gen;
    Rcpp::traits::input_parameter< ListOf<IntegerVector> >::type adj_list(adj_listSEXP);
    Rcpp::traits::input_parameter< int >::type k(kSEXP);
    Rcpp::traits::input_parameter< int >::type tabu_size(tabu_sizeSEXP);
    Rcpp::traits::input_parameter< int >::type rep(repSEXP);
    Rcpp::traits::input_parameter< int >::type nbmax(nbmaxSEXP);
    rcpp_result_gen = Rcpp::wrap(graph_coloring_tabucol(adj_list, k, tabu_size, rep, nbmax));
    return rcpp_result_gen;
END_RCPP
}

static const R_CallMethodDef CallEntries[] = {
    {"_graphcoloring_graph_coloring_dsatur", (DL_FUNC) &_graphcoloring_graph_coloring_dsatur, 1},
    {"_graphcoloring_graph_coloring_msc", (DL_FUNC) &_graphcoloring_graph_coloring_msc, 1},
    {"_graphcoloring_graph_coloring_lmxrlf", (DL_FUNC) &_graphcoloring_graph_coloring_lmxrlf, 1},
    {"_graphcoloring_graph_coloring_hybrid_dsatur_tabucol", (DL_FUNC) &_graphcoloring_graph_coloring_hybrid_dsatur_tabucol, 1},
    {"_graphcoloring_graph_coloring_hybrid_lmxrlf_tabucol", (DL_FUNC) &_graphcoloring_graph_coloring_hybrid_lmxrlf_tabucol, 1},
    {"_graphcoloring_graph_coloring_tabucol", (DL_FUNC) &_graphcoloring_graph_coloring_tabucol, 5},
    {NULL, NULL, 0}
};

RcppExport void R_init_graphcoloring(DllInfo *dll) {
    R_registerRoutines(dll, NULL, CallEntries, NULL, NULL);
    R_useDynamicSymbols(dll, FALSE);
}


src/graph_coloring.cpp
#include <Rcpp.h>

#include "GraphColoring/Header/coloring_algorithm.hpp"
#include "GraphColoring/Header/dsatur.hpp"
#include "GraphColoring/Header/mcs.hpp"
#include "GraphColoring/Header/lmxrlf.hpp"
#include "GraphColoring/Header/hybrid_dsatur.hpp"
#include "GraphColoring/Header/hybrid_lmxrlf.hpp"

using namespace Rcpp;

using std::vector;
using std::map;
using std::string;

using GraphColoring::Dsatur;
using GraphColoring::Mcs;
using GraphColoring::Lmxrlf;
using GraphColoring::HybridDsatur;
using GraphColoring::HybridLmxrlf;
using GraphColoring::GraphColor;

map<string, vector<string> > as_input_graph(ListOf<IntegerVector> adj_list) {
  map<string, vector<string> > input_graph;

  for(ListOf<IntegerVector>::iterator it = adj_list.begin(); it != adj_list.end(); ++it) {
    IntegerVector neighbors = as<IntegerVector>(*it);

    string node = std::to_string(it.index() + 1);
    input_graph[node] = as<vector<string> >(as<CharacterVector>(neighbors));
  }

  return input_graph;
}

IntegerVector as_coloring(GraphColor *graph, int n) {
  map<string,int> coloring = graph->color();

  IntegerVector output(n);
  for(int i = 0; i < n; ++i) {
    string node = std::to_string(i + 1);
    output(i) = coloring[node] + 1;
  }

  return output;
}


//' @describeIn graph_coloring Color graph using DSATUR algorithm
//' \insertCite{Brelaz:1979:NMC:359094.359101}{graphcoloring}
//' @export
// [[Rcpp::export]]
IntegerVector graph_coloring_dsatur(ListOf<IntegerVector> adj_list) {
  GraphColor *graph = new Dsatur(as_input_graph(adj_list));
  return as_coloring(graph, adj_list.size());
}

//' @describeIn graph_coloring Color graph using Maximum Cardinality Search(MCS) algorithm
//' \insertCite{Palsberg:2007:RAV:1273694.1273695}{graphcoloring}
//' @export
// [[Rcpp::export]]
IntegerVector graph_coloring_msc(ListOf<IntegerVector> adj_list) {
  GraphColor *graph = new Mcs(as_input_graph(adj_list));
  return as_coloring(graph, adj_list.size());
}

//' @describeIn graph_coloring Color graph using Least-constraining Most-constrained eXtended RLF(lmXRLF) algorithm
//' \insertCite{Kirovski:1998:ECL:277044.277165}{graphcoloring}
//' @export
// [[Rcpp::export]]
IntegerVector graph_coloring_lmxrlf(ListOf<IntegerVector> adj_list) {
  GraphColor *graph = new Lmxrlf(as_input_graph(adj_list));
  return as_coloring(graph, adj_list.size());
}

//' @describeIn graph_coloring Color graph using a hybrid of DASTUR and TabuCol algorithm
//' \insertCite{Kirovski:1998:ECL:277044.277165,Brelaz:1979:NMC:359094.359101,Hertz:1987:UTS:44141.44146}{graphcoloring}
//' @export
// [[Rcpp::export]]
IntegerVector graph_coloring_hybrid_dsatur_tabucol(ListOf<IntegerVector> adj_list) {
  GraphColor *graph = new HybridDsatur(as_input_graph(adj_list));
  return as_coloring(graph, adj_list.size());
}

//' @describeIn graph_coloring Color graph using a hybrid of lmXRLF and TabuCol algorithm
//' \insertCite{Kirovski:1998:ECL:277044.277165,Hertz:1987:UTS:44141.44146}{graphcoloring}
//' @export
// [[Rcpp::export]]
IntegerVector graph_coloring_hybrid_lmxrlf_tabucol(ListOf<IntegerVector> adj_list) {
  GraphColor *graph = new HybridLmxrlf(as_input_graph(adj_list));
  return as_coloring(graph, adj_list.size());
}

//' @describeIn graph_coloring Color graph using TabuCol algorithm
//' \insertCite{Hertz:1987:UTS:44141.44146}{graphcoloring}
//' @export
// [[Rcpp::export]]
IntegerVector graph_coloring_tabucol(ListOf<IntegerVector> adj_list, int k, int tabu_size = 25, int rep = 100, int nbmax = 1000) {
  GraphColor *graph = new Tabucol(as_input_graph(adj_list), k, tabu_size, rep, nbmax);
  IntegerVector coloring = as_coloring(graph, adj_list.size());

  if(!graph->is_valid()) {
    stop("Graph cannot be colored with " + std::to_string(k) + " colors!");
  }

  return coloring;
}


// You can include R code blocks in C++ files processed with sourceCpp
// (useful for testing and development). The R code will be automatically
// run after the compilation.
//

/*** R
timesTwo(42)
*/


tools/config/configure.R
# Prepare your package for installation here.
# Use 'define()' to define configuration variables.
# Use 'configure_file()' to substitute configuration values.

dir.create("src/GraphColoring", showWarnings = FALSE)
file.copy("src/GraphColoring-raw/Header", "src/GraphColoring", recursive = TRUE)
file.copy("src/GraphColoring-raw/Source", "src/GraphColoring", recursive = TRUE)
unlink("src/GraphColoring/Source/main.cpp")
unlink("src/GraphColoring/Source/test.cpp")

cpp_files <- list.files(
  "src/GraphColoring",
  "\\.cpp$",
  full.names = TRUE,
  recursive = TRUE
)
for (file in cpp_files) {
  lines <- readLines(file, warn = FALSE)

  lines <- gsub("using std::cout;", "", lines, fixed = TRUE)
  lines <- gsub("using std::cerr;", "", lines, fixed = TRUE)

  lines <- gsub("cout", "Rcout", lines, fixed = TRUE)
  lines <- gsub("cerr", "Rcerr", lines, fixed = TRUE)

  lines <- gsub("std::Rcout", "Rcout", lines, fixed = TRUE)
  lines <- gsub("std::Rcerr", "Rcerr", lines, fixed = TRUE)

  lines <- gsub("srand(time(NULL));", "", lines, fixed = TRUE)
  lines <- gsub("rand\\(\\) % ([A-Za-z->_.]+\\(\\))", "Rcpp::sample(\\1, 1)[0] - 1", lines, perl = TRUE)
  lines <- gsub("rand\\(\\) % ([A-za-z->]+)", "Rcpp::sample(\\1, 1)[0] - 1", lines, perl = TRUE)

  lines <- c(
    "#include <Rcpp.h>",
    "using Rcpp::Rcout;",
    "using Rcpp::Rcerr;",
    lines,
    ""
  )

  writeLines(lines, file)
}

h_files <- list.files(
  "src/GraphColoring",
  "\\.h$",
  full.names = TRUE,
  recursive = TRUE
)
for (file in h_files) {
  lines <- readLines(file, warn = FALSE)

  lines <- c(
    lines,
    ""
  )

  writeLines(lines, file)
}


R/color.R
#' Color nodes using Graph Coloring Algorithm
#'
#' These functions are [`tidygraph`] wrapper around the various [graph coloring algorithms][graph_coloring()].
#' They automatically use the graph that is being computed on, and
#' otherwise passes on its arguments to the relevant coloring function. The return value is always
#' a integer vector of assigned color index so that neighboring nodes never share the same color.
#'
#' @name color_graph
#' @rdname color_graph
#'
#' @param k number of colors to use for graph coloring
#' @param tabu_size size of tabu list
#' @param rep number of inner iterations
#' @param nbmax maximum number of non-improving outer iterations
#'
#' @examples
#' library(tidygraph)
#'
#' if (requireNamespace("ggraph", quietly = TRUE)) {
#'   library(ggraph)
#'   set.seed(42)
#'
#'   play_islands(5, 10, 0.8, 3) %>%
#'     mutate(color = as.factor(color_dsatur())) %>%
#'     ggraph(layout = 'kk') +
#'     geom_edge_link(aes(alpha = ..index..), show.legend = FALSE) +
#'     geom_node_point(aes(color = color), size = 7) +
#'     theme_graph("")
#' }
NULL

#' Graph Coloring over Adjacency List
#'
#' @description
#' These functions perform graph coloring using various algorithms over an adjacency list.
#'
#' In graph theory, graph coloring is a special case of graph labeling;
#' it is an assignment of labels traditionally called "colors" to elements of a graph subject
#' to certain constraints. In its simplest form, it is a way of coloring the vertices of a graph
#' such that no two adjacent vertices share the same color; this is called a vertex coloring.
#'
#' @details
#'
#' [graph_coloring_hybrid_dsatur_tabucol()] and [graph_coloring_hybrid_lmxrlf_tabucol()] use a hybrid approach
#' to run DSATUR and lmXRLF first to determine an upper bound for the graph chromatic number. It then searches
#' better solutions by running lowered chromatic number through TabuCol to check if the graph can be colored
#' with less colors.
#'
#' @param adj_list an adjacency list in the format of `list` of `integer` vector. The outer list
#' should enumerate nodes comprehensively and each integer vector enumerates corresponding neighboring nodes
#'
#' @inheritParams color_graph
#'
#' @references
#' <https://en.wikipedia.org/wiki/Graph_coloring>
#'
#' <https://github.com/brrcrites/GraphColoring>
#'
#' \insertRef{Brelaz:1979:NMC:359094.359101}{graphcoloring}
#'
#' \insertRef{Palsberg:2007:RAV:1273694.1273695}{graphcoloring}
#'
#' \insertRef{Kirovski:1998:ECL:277044.277165}{graphcoloring}
#'
#' \insertRef{Hertz:1987:UTS:44141.44146}{graphcoloring}
#'
#' @name graph_coloring
#' @rdname graph_coloring
#' @seealso [color_graph()]
#' @examples
#' library(tidygraph)
#'
#' if (requireNamespace("sf", quietly = TRUE) && requireNamespace("USAboundaries", quietly = TRUE)) {
#'  library(sf)
#'  library(USAboundaries)
#'
#'  us_states() %>%
#'    filter(!(name %in% c("Alaska", "District of Columbia", "Hawaii", "Puerto Rico"))) %>%
#'    transmute(
#'      color = st_intersects(.) %>%
#'        graph_coloring_dsatur() %>%
#'        as.factor()
#'      ) %>%
#'    plot()
#' }
NULL

color_with <- function(f, ...) {
  expect_nodes()

  graph <- tidygraph::.G()
  adj_list <- igraph::as_adj_list(graph)
  f(adj_list, ...)
}

#' @describeIn color_graph Color graph using [graph_coloring_dsatur()]
#' @export
color_dsatur <- function() {
  color_with(graph_coloring_dsatur)
}

#' @describeIn color_graph Color graph using [graph_coloring_msc()]
#' @export
color_msc <- function() {
  color_with(graph_coloring_msc)
}

#' @describeIn color_graph Color graph using [graph_coloring_lmxrlf()]
#'
#' **WARNING** Algorithm is unstable and requires additional testing
#' @export
color_lmxrlf <- function() {
  color_with(graph_coloring_lmxrlf)
}

#' @describeIn color_graph Color graph using [graph_coloring_hybrid_lmxrlf_tabucol()]
#'
#' **WARNING** Algorithm is unstable and requires additional testing
#' @export
color_hybrid_lmxrlf_tabucol <- function() {
  color_with(graph_coloring_hybrid_lmxrlf_tabucol)
}

#' @describeIn color_graph Color graph using [graph_coloring_hybrid_dsatur_tabucol()]
#' @export
color_hybrid_dsatur_tabucol <- function() {
  color_with(graph_coloring_hybrid_dsatur_tabucol)
}

#' @describeIn color_graph Color graph using [graph_coloring_tabucol()]
#' @export
color_tabucol <- function(k, tabu_size = 25 , rep = 100, nbmax = 1000) {
  color_with(graph_coloring_tabucol, k, tabu_size, rep, nbmax)
}
